### PR TITLE
Add percentage size unit for Raster Image Marker and Raster fill layers symbology.

### PR DESCRIPTION
**Pull request by beketata**
## Description

According to the [Feature Request](https://github.com/qgis/QGIS/issues/30057) this patch adds new percentage (percents of original image size) value of size units for **Raster Image Marker** and **Raster fill** (by [nirvn](https://github.com/qgis/QGIS/pull/34869#issuecomment-595090465) suggestion) layers.

For example a layer contains raster images of various sizes and aspect ratios:

crocodile giraffe

If we use an existing Raster image marker **size unit** values the layer will look like this (with normal croc, but huge giraffe):

crocGir

or like this (with normal giraffe but tiny croc):

GirCroc

The new **percentage** value of size units is intended to fix this issue.

===========================================

Both unit tests "qgis_rasterfilltest" and "qgis_rastermarkertest" were rewritten with a set of new control images for testing proposed **percentage** value of size units.

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

### CODE REVIEW

**Comment by beketata** [comment]
How to know which unsuccessful tests require my participation and which do not?

File: src/core/symbology/qgsfillsymbollayer.cpp [hunk 1]
```diff
@@ -3906,6 +3906,21 @@ void QgsRasterFillSymbolLayer::applyPattern( QBrush &brush, const QString &image
   if ( width > 0 )
   {
     size.setWidth( context.renderContext().convertToPainterUnits( width, mWidthUnit, mWidthMapUnitScale ) );
+
+    // RenderPercentage Unit Type takes original image size
+    if ( mWidthUnit == QgsUnitTypes::RenderPercentage )
```
Comment (nyalldawson) [inline src/core/symbology/qgsfillsymbollayer.cpp:3911]: this should be in an "else" block, avoiding the unused setWidth call from above. E.g.

    if ( mWidthUnit != QgsUnitTypes::RenderPercentage )
    {
      size.setWidth( context.renderContext().convertToPainterUnits( width, mWidthUnit, mWidthMapUnitScale ) );
    }
    else
    {
       QSize originalSize = QgsApplication::imageCache()->originalSize( imageFilePath );
       ...
    }

File: tests/src/core/testqgsrasterfill.cpp [hunk 2]
```diff
@@ -96,11 +100,18 @@ void TestQgsRasterFill::initTestCase()
...
-  //setup raster fill
+  // Setup raster fill
   mRasterFill = new QgsRasterFillSymbolLayer();
+  mRasterFill->setImageFilePath( mTestDataDir + QStringLiteral( "sample_image.png" ) );
+  mRasterFill->setWidthUnit( QgsUnitTypes::RenderPixels );
+  mRasterFill->setWidth( 30.0 );
+  mRasterFill->setCoordinateMode( QgsRasterFillSymbolLayer::Feature );
+  mRasterFill->setOpacity( 1.0 );
+  mRasterFill->setOffset( QPointF( 0, 0 ) );
+  mRasterFill->setOffsetUnit( QgsUnitTypes::RenderPixels );
```
Comment (nyalldawson) [inline tests/src/core/testqgsrasterfill.cpp:113]: Can you leave out these changes? The intention with putting them in the ::init() method is that that method gets called before every test, so if an individual test messes with the mRasterFill definition then any changes it makes are reset when the next test method is called.

File: tests/src/core/testqgsrastermarker.cpp [hunk 3]
```diff
@@ -68,35 +76,44 @@ class TestQgsRasterMarker : public QObject
...
 
-  //setup the raster marker symbol
+  // Setup the raster marker symbol
   mRasterMarker = new QgsRasterMarkerSymbolLayer();
+  mRasterMarker->setPath( mTestDataDir + QStringLiteral( "sample_image.png" ) );
+  mRasterMarker->setSizeUnit( QgsUnitTypes::RenderPixels );
+  mRasterMarker->setSize( 30.0 );
+  mRasterMarker->setHorizontalAnchorPoint( QgsMarkerSymbolLayer::HCenter );
+  mRasterMarker->setVerticalAnchorPoint( QgsMarkerSymbolLayer::VCenter );
+  mRasterMarker->setOffsetUnit( QgsUnitTypes::RenderPixels );
```
Comment (nyalldawson) [inline tests/src/core/testqgsrastermarker.cpp:112]: Same here -- the original version was more robust to failing tests

**Review by nyalldawson — COMMENTED** [review]
The symbology changes look good!

You'll need to revert most of the changes to the test suite though -- as noted above, they cause the tests to become less robust, and have actually broken some of the existing tests (as you can see by the need to change the reference images. You shouldn't have to change any of the existing reference images in this PR).

**Comment by nyalldawson** [comment]
Nice work @beketata , and sorry about the delayed review! If you can fix the tests as noted above then this is good to merge.

File: tests/src/core/testqgsrasterfill.cpp (hunk 2 above)
Comment (beketata) [reply tests/src/core/testqgsrasterfill.cpp:113]: Do you mean that each new test should start with exactly the same source image defined in the ::init() ?

File: src/core/symbology/qgsfillsymbollayer.cpp (hunk 1 above)
Comment (beketata) [reply src/core/symbology/qgsfillsymbollayer.cpp]: Did it.

File: tests/src/core/testqgsrasterfill.cpp (hunk 2 above)
Comment (nyalldawson) [reply tests/src/core/testqgsrasterfill.cpp:113]: It should start with new instances of the symbols. Just revert your changes, add the new test alone, and it should all work fine 😁

Comment (beketata) [reply tests/src/core/testqgsrasterfill.cpp]: Done.

File: tests/src/core/testqgsrastermarker.cpp (hunk 3 above)
Comment (beketata) [reply tests/src/core/testqgsrastermarker.cpp]: Done.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Nailed it -- thanks!

**Pull request merged**